### PR TITLE
Make out_identifier id and namespace optional

### PR DESCRIPTION
**Pull request by ranveernitk**

### CODE REVIEW

File: lib/rule-interface/rule_executor.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 module RuleInterface
   module RuleExecutor
-    def execute!(data_hash:, namespace:, container:, package:)
-      payload = Converter.hash_to_drool(data_hash: data_hash, namespace: namespace, package: package)
+    def execute!(data_hash:, namespace: nil, container:, package:, sesson: 'sesson')
```
Comment (vishalvijay) [inline lib/rule-interface/rule_executor.rb]: Move mandatory value to the beginning `container `

File: lib/rule-interface/converter.rb [hunk 2]
```diff
@@ -2,18 +2,22 @@ module RuleInterface
...
     extend self
 
-    def hash_to_drool(data_hash:, namespace:, package:)
+    def hash_to_drool(data_hash:, namespace:, package:, session:)
       payload = {}
-      payload[:lookup] = 'session'
+      payload[:lookup] = session
       payload[:commands] = []
 
+      Thread.current[:id] = 1
```
Comment (vishalvijay) [inline lib/rule-interface/converter.rb]: Move `:id` to a constant and rename the value to `:rule_interface_out_identifier_id`

File: lib/rule-interface/converter.rb [hunk 3]
```diff
@@ -55,7 +55,14 @@ def insert_object(data_object:, object_type:, package:, return_object: true)
       insert_object = {}
       insert_object[:insert] = {}
       insert_object[:insert]['return-object'] = return_object
-      insert_object[:insert]['out-identifier'] = "#{object_type}##{data_object[:id]}" if return_object
+
+      out_identifier_id = data_object[:id]
+      unless out_identifier_id
```
Comment (vishalvijay) [inline lib/rule-interface/converter.rb]: `if out_identifier_id.blank?`

**Review by vishalvijay — CHANGES_REQUESTED** [review]
- Specs
- Readme

File: lib/rule-interface/converter.rb [hunk 4]
```diff
@@ -2,18 +2,22 @@ module RuleInterface
...
+        payload[:commands] << insert_object(
+          data_object: {name: namespace},
+          object_type: 'namespace',
+          package: package,
+          return_object: false
+        )
+      end
 
       data_hash.each do |object_type, data_objects|
         data_objects.each do |data_object|
```
Comment (shivanshgaur) [inline lib/rule-interface/converter.rb]: Should we have a check here for the case where user wants to insert single object. Something like following:
`
data_hash: {
    product: {
        id: 12,
        name: 'Blah'
    }
}
`

Here product is just a hash and not array.

I suggest to use
`data_objects = [data_objects] unless data_objects.is_a? Array`

**Review by shivanshgaur — DISMISSED** [review]

**Review by vishalvijay — APPROVED** [review]

**Pull request merged**